design/user_controller_workspaces.md

```diff
@@ -0,0 +1,164 @@
+
+Writes just append the path to be written to the workspace path: `${workspaceAddr}/${writePath}`
+Deletes are also writes, but of a sentinel tombstone value.
+Reads are now a getObject/listObjects call on that prefix, implementing our "DiffIterator".
+
+When calling `commit`, we take that `diffIterator` and apply it. Once we have a commit ID, we do a `CAS` on the ref: if no one else committed so far, this completes the commit process successfully. Otherwise, it's an `ErrConcurrentCommit` and should be retried/aborted.
+Reading these refs is done using the same batching method we currently do for PG, so we could support pretty much ANY external storage for this as long as it is:
+
+* Read-after-write consistent
+* Has at least one of `{CAS, conditional-write, serializable transaction}`
+
+On this list are: DynamoDB (which we can abstract away(!)), PostgreSQL, MySQL, MongoDB, even a Hive Metastore..
+Using good object stores (i.e., practically anything other than S3), this could be done with no DB at all, with some additional latency.
+
+It is the user's responsibility to ensure they commit at the end of their job/pipeline/script,
+but it's possible to pass around that `workspaceID` between jobs and processes, even if reading/writing concurrently.
+
+Once a commit is done for a given workspace, it should no longer be used. if using the lakeFS SDK (as shown above),
+calling any operation on a `commit()`ed `Workspace` object will raise an error. We can go further and place a "lockfile" (`_lakefs_committed/${workspaceID}`) and check it before writing if we think this is necessary (on a best effort basis).
+
+## Using the S3 gateway
+
+just append a workspace ID: `s3://my-repo/master+abc123/my/collection/` to read/write.
+otherwise, just pass a ref as you normally would, to read committed data.
+
+## Out-of-band ingest
+
+Now that a commit is basically a diffIterator on top of `listObjects` we can do something similar for stream processing.
+Working with streams is different: we can't always ensure no writes are happening, but fortunately, most streaming processors 
+do [guarantee immutability](https://stackoverflow.com/a/61201083): A written file is never overwritten, only written once and potentially deleted: this allows us to take
+a regular old object store path and do:
+
+```sh
+$ lakectl ingest commit \
+    --from-path "s3://bucket/kafka/events/2021/" \
+    --to-path "lakefs://my-repo@ingest-branch/kafka/events/2021/" \
+    -m "ingest events: $(date -u)"
+
+# Or just see what changes would be introduced
+$ lakectl ingest diff \
+    --from-path "s3://bucket/kafka/events/2021/" \
+    --to-path "lakefs://my-repo@ingest-branch/kafka/events/2021/"
+```
+Here, instead of treating the given object store location as a diffIterator, we look at it as a subtree, to be diffed with the underlying commit.
+
+
+## IAM permissions
+
+set them normally on `s3://storagenamespace/*/path/a`, `s3://storagenamespace/*/path/b`, etc.
```

Comment:
If the workspace token is generated with the branch in mind, we might be able to do something with branch-based permission too.